### PR TITLE
available_variables and available_attributes for the C API

### PR DESCRIPTION
**Pull request by dmitry-ganyushin**

### CODE REVIEW

File: bindings/C/adios2/c/adios2_c_io.cpp [hunk 1]
```diff
@@ -874,6 +874,60 @@ adios2_error adios2_remove_all_attributes(adios2_io *io)
...
+            io, "for adios2_io, in call to adios2_available_variables");
+        std::map<std::string, std::map<std::string, std::string>> varInfo =
+            reinterpret_cast<adios2::core::IO *>(io)->GetAvailableVariables();
+        *size = varInfo.size();
+        char **names = (char **)malloc(*size * sizeof(char *));
+
+        size_t cnt = 0;
+        for (auto var : varInfo)
+        {
+            int len = var.first.length();
```
Comment (williamfgc) [inline bindings/C/adios2/c/adios2_c_io.cpp]: `len` should stay as `std::size_t` as returned by `length()` and the input type to `malloc`

**Review by williamfgc — COMMENTED** [review]
@dmitry-ganyushin Great work. It should be documented that the returned `char**` must be manually deallocated by the user. I remember @pnorbert needed something similar for the Matlab bindings that use the C APIs.

File: bindings/C/adios2/c/adios2_c_io.cpp (hunk 1 above)
Comment (dmitry-ganyushin) [reply bindings/C/adios2/c/adios2_c_io.cpp]: Thanks a lot! The main question is whether we should have unicode strings or not.

Comment (williamfgc) [reply bindings/C/adios2/c/adios2_c_io.cpp]: If I remember correctly unicode was supported as-is in the C++ `std::string` side. You can check with @pnorbert .

**Pull request merged**